src/accountsdb/cache.zig

```diff
@@ -0,0 +1,331 @@
+    const allocator = std.testing.allocator;
+    var random = std.rand.DefaultPrng.init(19);
+    const rng = random.random();
+
+    var accounts_cache = try AccountsCache.init(allocator, 10, 1);
+    defer accounts_cache.deinit();
+
+    const pubkey = Pubkey.random(rng);
+    const slot = 1;
+
+    const result = accounts_cache.get(slot, pubkey);
+    try std.testing.expect(result == null);
+}
+
+test "AccountsCache put & copySlot ref counting" {
+    const allocator = std.testing.allocator;
+    var random = std.rand.DefaultPrng.init(19);
+    const rng = random.random();
+
+    var accounts_cache = try AccountsCache.init(allocator, 10, 2);
+    defer accounts_cache.deinit();
+
+    const account = try Account.random(allocator, rng, 1);
+    defer account.deinit(allocator);
+
+    const pubkey = Pubkey.random(rng);
+    const old_slot = 1;
+    const new_slot = 2;
+
+    try accounts_cache.put(old_slot, pubkey, account);
+
+    try std.testing.expectEqual(accounts_cache.get(old_slot, pubkey).?.ref_count.load(.acquire), 1);
+
+    try accounts_cache.shiftLRUForward(old_slot, new_slot);
+
+    try std.testing.expectEqual(accounts_cache.get(old_slot, pubkey).?.ref_count.load(.acquire), 2);
+
+    const cached_account = accounts_cache.get(new_slot, pubkey);
+    try std.testing.expect(cached_account != null);
+}
+
+test "AccountsCache max slots" {
+    const allocator = std.testing.allocator;
+    var random = std.rand.DefaultPrng.init(19);
+    const rng = random.random();
+
+    var accounts_cache = try AccountsCache.init(allocator, 10, 2);
+    defer accounts_cache.deinit();
+
+    const account = try Account.random(allocator, rng, 1);
+    defer account.deinit(allocator);
+
+    const pubkey = Pubkey.random(rng);
+
+    try accounts_cache.put(1, pubkey, account);
+    try std.testing.expect(accounts_cache.get(1, pubkey) != null);
+    try accounts_cache.shiftLRUForward(1, 2);
+    try std.testing.expect(accounts_cache.get(1, pubkey) != null);
+    // create 3rd slot, max slots = 2, 1st slot evicted
+    try accounts_cache.shiftLRUForward(2, 3);
+    try std.testing.expect(accounts_cache.get(1, pubkey) == null);
+    try std.testing.expect(accounts_cache.slot_lrus.count() == 2);
+}
+
+// test "AccountsCache put returns error on duplicate" {
+//     const allocator = std.testing.allocator;
+//     var random = std.rand.DefaultPrng.init(19);
+//     const rng = random.random();
+
+//     const account = try Account.random(allocator, rng, 1);
+//     defer account.deinit(allocator);
+
+//     var accounts_cache = try AccountsCache.init(allocator, 10, 1);
+//     defer accounts_cache.deinit();
+
+//     const pubkey = Pubkey.random(rng);
+//     const slot = 1;
+
+//     try accounts_cache.put(slot, pubkey, account);
+
+//     // Trying to insert the same account again should fail
+//     try std.testing.expectEqual(error.AlreadyExistsInCache, accounts_cache.put(slot, pubkey, account));
+// }
```

Comment:
There's a memory leak hiding here somewhere, will re-enable this test when found

Reply:
Fixed - slot_lru.put was overwriting the old value in cases where the key already existed, causing the value to be dangling and not freed. Fixed by introducing putNoClobber.